opencl-aot/source/main.cpp

```diff
@@ -0,0 +1,488 @@
+    if (clFailed(CLErr)) {
+      std::cerr << formatCLError("Failed to release OpenCL context", CLErr);
+    }
+  }
+};
+
+using CLContextUPtr = std::unique_ptr<
+    _cl_context, decltype(ContextDeleter)>; // decltype(CLContextUPtr.get()) is
+                                            // same as cl_context, should
+                                            // initialized with ContextDeleter
+
+/*! \fn std::tuple<std::vector<CLProgramUPtr>, std::string, cl_int>
+ * generateProgramsFromBinaries(const std::vector<std::string> &FileNames,
+ * cl_platform_id PlatformId, cl_device_id DeviceId)
+ * \brief Generate OpenCL program from OpenCL program binary (in ELF format) or
+ * SPIR-V binary file
+ * \param FileNames (const std::vector<std::string>).
+ * \param PlatformId (cl_platform_id).
+ * \param DeviceId (cl_device_id).
+ * \return a tuple of vector of unique pointers to programs, error message and
+ * return code.
+ */
+std::tuple<std::vector<CLProgramUPtr>, std::string, cl_int>
+generateProgramsFromBinaries(const std::vector<std::string> &FileNames,
+                             cl_platform_id PlatformId, cl_device_id DeviceId) {
+  // step 0: define internal types
+
+  enum SupportedTypes : int8_t { BEGIN, ELF = BEGIN, SPIRV, UNKNOWN, END };
+  static std::map<SupportedTypes, const std::string> SupportedTypesToNames{
+      {ELF, "OpenCL program binary"}, {SPIRV, "SPIR-V"}, {UNKNOWN, "UNKNOWN"}};
+  assert(SupportedTypesToNames.size() == SupportedTypes::END &&
+         "unexpected name"); // to prevent incorrect refactoring
+
+  enum { FTYPE, DATA };
+
+  using Content = std::pair<SupportedTypes, std::vector<char>>;
+
+  // step 1: validate input
+
+  if (FileNames.empty() ||
+      std::any_of(FileNames.begin(), FileNames.end(),
+                  [](const std::string &File) { return File.empty(); })) {
+    return std::make_tuple(
+        std::vector<CLProgramUPtr>(),
+        "List of input binaries is empty or contains empty filename\n",
+        OPENCL_AOT_LIST_OF_INPUT_BINARIES_IS_EMPTY);
+  }
+
+  assert(PlatformId != nullptr && "Platform must not be empty");
+  assert(DeviceId != nullptr && "Device must not be empty");
+
+  // step 2: read files and put them to cache
+
+  std::unordered_map<std::string, Content> FileNameToContentMap;
+  enum { FNAME, CONTENT };
+  FileNameToContentMap.reserve(FileNames.size());
+
+  cl_int CLErr(CL_SUCCESS);
+  std::string ErrorMessage;
+  for (const std::string &FileName : FileNames) {
+    std::vector<char> FileContent;
+    std::tie(FileContent, ErrorMessage, CLErr) = readBinaryFile(FileName);
+
+    if (clFailed(CLErr)) {
+      return std::make_tuple(std::vector<CLProgramUPtr>(), ErrorMessage, CLErr);
+    }
+
+    FileNameToContentMap.emplace(
+        FileName,
+        std::make_pair(SupportedTypes::UNKNOWN, std::move(FileContent)));
+  }
+
+  // step 3: find out type of each file
+
+  for (auto &Value : FileNameToContentMap) {
+    auto &FileContent = std::get<CONTENT>(Value);
```

Comment:
FYI: [Don't "almost always" use `auto`](https://llvm.org/docs/CodingStandards.html#use-auto-type-deduction-to-make-code-more-readable)

This particular case is debatable, since it is possible to manually deduce the type of `FileContext`, for example, by scrolling the code a bit (within the current function scope). But anyway, it requires 4 manual steps: `FileContent` -> `Value` -> `FileNameToContentMap` -> `Content`

Reply:
I see your concern. I replaced ```auto``` with ```Content``` for such case. After that the deduction of ```FileType``` and ```BinaryData``` will be easier